docs/design/secrets.md

```diff
@@ -0,0 +1,187 @@
+provides a mysql database needs to provide the username, password, and database name to consumers
+so that they can authenticate and use the correct database. Pods consuming the mysql service would
+also consume the secrets associated with the mysql service.
+
+## Consuming secrets as environment variables
+
+Some images will expect to receive configuration items as environment variables instead of files.
+We should consider what the best way to allow this is; there are a few different options:
+
+1.  Force the user to adapt files into environment variables.  Users can store secrets that need to
+    be presented as environment variables in a format that is easy to consume from a shell:
+
+        $ cat /etc/secrets/my-secret.txt
+        export MY_SECRET_ENV=MY_SECRET_VALUE
+
+    The user could `source` the file at `/etc/secrets/my-secret` prior to executing the command for
+    the image either inline in the command or in an init script, 
+
+2.  Give secrets an attribute that allows users to express the intent that the platform should
+    generate the above syntax in the file used to present a secret.  The user could consume these
+    files in the same manner as the above option.
+
+3.  Give secrets attributes that allow the user to express that the secret should be presented to
+    the container as an environment variable.  The container's environment would contain the
+    desired values and the software in the container could use them without accomodation the
+    command or setup script.
+
+**TODO**: Detailed assessment of the above options
+
+## Analysis TODOs
+
+Collecting remaining TODOs here:
+
+1.  Describe the lifecycle of secrets
+2.  Determine how/whether secrets interact with service accounts; are namespaces enough for the
+    security scope? Can/should my security context affect what secrets I have access to?
+3.  How will pods know when secret values change?  Should there be a way to express that a pod
+    which consumes a secret should be restarted when the secret changes?
+
+## Proposed Design
+
+### Overview
+
+This design proposes mounting secrets with a new volume type. The secrets volume will be backed
+by a volume plugin that does the actual work of fetching the secret and placing it on the
+filesystem to be mounted in the container. It should be possible to mount single files as part of a
+secret. Secrets may consist of multiple files. For example, an SSH key pair. In order to remove the
+burden from the end user in specifying every file that a secret consists of, it should be possible
+to mount all files provided by a secret with a single ```VolumeMount``` entry in the container
+specification.
+
+### Secret Volume Source
+
+A new Secret type of volume source will be added to the ```VolumeSource``` struct in the API:
+
+```go
+type VolumeSource struct {
+     ... 
+     Secret *Secret `json:"secret"`
+}
+
+type Secret struct {
+     // Ref is a reference to the actual secret
+     Ref string 
```

Comment:
If secrets are stored as a kube resource, then this should be an ObjectReference. I guess the problem is if you choose a different place to store them (say a docker registry), then the reference would be to an image. I think not committing to a specific way of storing secrets doesn't help here. Maybe we should just say secrets are a Kube resource and see where that goes.